### PR TITLE
Dataset unification

### PR DESCRIPTION
**Pull request by aibobrov**

### CODE REVIEW

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/CompositeTransformation.kt [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.logging.Logger
+
+/**
+ * Transformation that run all inner transformation until nothing is changed
+ */
+object CompositeTransformation : Transformation() {
+    private val LOG = Logger.getLogger(javaClass.name)
+
+    override val key: String = "CompositeTransformation"
+    private val transformations = arrayListOf(
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/CompositeTransformation.kt:26]: What do you think about using Reflektion here? I know that is is not fast, but I think for our goals it is ok. In addition, in a couple of weeks we can use compile-time reflection as soon as the [Reflekt](https://github.com/JetBrains-Research/reflekt) is published

Comment (nbirillo) [reply src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/CompositeTransformation.kt:26]: I am not sure I am right, we can discuss it

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/CompositeTransformation.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                    """Transformation error {$e}: 
+                        |Previous Code=${previousTree.text}
+                        |Current Code=${psiTree.text}
+                        |""".trimMargin()
+                }
+                throw e
+            }
+
+            LOG.info { "Previous text[$iterationNumber]:\n${previousTree.text}\n" }
+            LOG.info { "Current text[$iterationNumber]:\n${psiTree.text}\n\n" }
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/CompositeTransformation.kt]: I think it is DEBUG level

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/CompositeTransformation.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        } while (!previousTree.textMatches(psiTree.text))
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/CompositeTransformation.kt]: Should we forcefully stop after say 100 iterations or some other value?

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnification.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import org.jetbrains.research.ml.coding.assistant.dataset.model.DatasetRecord
+import org.jetbrains.research.ml.coding.assistant.dataset.model.DynamicSolution
+import org.jetbrains.research.ml.coding.assistant.solutionSpace.utils.psiCreator.PsiCreator
+import org.jetbrains.research.ml.coding.assistant.unification.model.DatasetPartialSolution
+import org.jetbrains.research.ml.coding.assistant.unification.model.DynamicIntermediateSolution
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.logging.Logger
+
+@Service
+interface DatasetUnification {
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnification.kt:20]: Please add general description for this interface

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnification.kt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    /**
+     * Unifies the given solution dynamic
+     * @param dynamicSolution raw dataset dynamic solution
+     * @return unified solution using ast transformation
+     */
+    fun transform(dynamicSolution: DynamicSolution): DynamicIntermediateSolution
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnification.kt]: Probably a better name is `unify` or something like this?

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnification.kt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+                unifyRecord(it)
+            }
+            .toList()
+    }
+
+    private fun unifyRecord(datasetRecord: DatasetRecord): DatasetPartialSolution {
+        val psiFile = fileFactory.initFileToPsi(datasetRecord.fragment)
+        WriteCommandAction.runWriteCommandAction(project) { // reformat the expected file
+            codeStyleManager.reformat(psiFile)
+        }
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnification.kt]: I think we should create a function for it. Also we can use this function in the tests

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/model/DatasetPartialSolution.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
+package org.jetbrains.research.ml.coding.assistant.unification.model
+
+import com.intellij.openapi.application.ApplicationManager
+import org.jetbrains.research.ml.coding.assistant.dataset.model.MetaInfo
+import org.jetbrains.research.ml.coding.assistant.solutionSpace.utils.psiCreator.PsiFileWrapper
+
+typealias DynamicIntermediateSolution = List<DatasetPartialSolution>
+
+/**
+ * Dataset partial solution
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/model/DatasetPartialSolution.kt]: Please, add more detailed info

File: src/test/kotlin/org/jetbrains/research/ml/coding/assistant/compositeTransformation/CompositeTransformationTest.kt [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import org.jetbrains.research.ml.coding.assistant.solutionSpace.utils.psiCreator.PsiCreator
+import org.jetbrains.research.ml.coding.assistant.unification.CompositeTransformation
+import org.jetbrains.research.ml.coding.assistant.util.ParametrizedBaseWithSdkTest
+import org.junit.Ignore
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+import java.io.File
+
+@Ignore
```
Comment (nbirillo) [inline src/test/kotlin/org/jetbrains/research/ml/coding/assistant/compositeTransformation/CompositeTransformationTest.kt]: Why should we ignore this test?

File: src/test/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnificationTest.kt [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    @JvmField
+    @Parameterized.Parameter(0)
+    var inFile: String? = null
+
+    @JvmField
+    @Parameterized.Parameter(1)
+    var outFile: String? = null
+
+    @Test
+    fun testBasic() {
```
Comment (nbirillo) [inline src/test/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnificationTest.kt]: Should we assert something? or at least check that there were no errors on run

File: src/test/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnificationTest.kt [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+            val graph = builder.build { CustomEdgeWeightCalculator(it) }
+            val imgFile = File("${taskSolutions.datasetTask.taskName}_graph.png").apply { createNewFile() }
+            val image = graph.generateImage()
+            ImageIO.write(image, "PNG", imgFile)
+        }
+    }
+
+    @Test
+    fun testStuff() {
```
Comment (nbirillo) [inline src/test/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnificationTest.kt]: Very strange name

File: src/test/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnificationTest.kt [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+        val inputDir = "/Users/artembobrov/Documents/masters/ast-transform/python/max_digit/"
```
Comment (nbirillo) [inline src/test/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnificationTest.kt]: Please, don't use these names

File: src/test/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnificationTest.kt [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+    override fun getTestDataPath(): String = getTmpProjectDir(toCreateFolder = false)
+
+    companion object {
+        lateinit var taskTrackerDataset: Dataset
+
+        @BeforeClass
+        @JvmStatic
+        fun setup() {
+            val path = "/Users/artembobrov/Documents/masters/ast-transform/python"
```
Comment (nbirillo) [inline src/test/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnificationTest.kt]: Absolute path

File: src/test/kotlin/org/jetbrains/research/ml/coding/assistant/compositeTransformation/CompositeTransformationTest.kt [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+@RunWith(Parameterized::class)
+class CompositeTransformationTest : ParametrizedBaseWithSdkTest(getTmpProjectDir(true)) {
```
Comment (nbirillo) [inline src/test/kotlin/org/jetbrains/research/ml/coding/assistant/compositeTransformation/CompositeTransformationTest.kt:16]: Please, add more test cases

**Review by nbirillo — CHANGES_REQUESTED** [review]
Great

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/CompositeTransformation.kt (hunk 1 above)
Comment (aibobrov) [reply src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/CompositeTransformation.kt:26]: mb now it's ok
but back in the days when anonymization problems had been destroying my evenings it wasn't the solution.

Comment (aibobrov) [reply src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/CompositeTransformation.kt:26]: but I personally do not see anything bad in this list.

Comment (elena-lyulina) [reply src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/CompositeTransformation.kt:26]: let's keep it for now and change it if there is extra time left

Comment (elena-lyulina) [reply src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/CompositeTransformation.kt:26]: also, all inner transformation -> all inner transformation**s**

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnification.kt [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.command.WriteCommandAction
+import com.intellij.openapi.components.Service
+import com.intellij.openapi.components.service
+import com.intellij.openapi.project.Project
+import com.intellij.psi.codeStyle.CodeStyleManager
+import org.jetbrains.research.ml.coding.assistant.dataset.model.DatasetRecord
+import org.jetbrains.research.ml.coding.assistant.dataset.model.DynamicSolution
+import org.jetbrains.research.ml.coding.assistant.solutionSpace.utils.psiCreator.PsiCreator
```
Comment (elena-lyulina) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/DatasetUnification.kt:9]: does solutionSpace package persist in this PR? I don't see it for some reason and my intelliJ cannot find PsiCreator

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/model/DatasetPartialSolution.kt (hunk 7 above)
Comment (elena-lyulina) [reply src/main/kotlin/org/jetbrains/research/ml/coding/assistant/unification/model/DatasetPartialSolution.kt]: Yeah, it'd be great to define partial solution

**Review by elena-lyulina — CHANGES_REQUESTED** [review]

**Review by elena-lyulina — APPROVED** [review]
Well done!

**Pull request merged**